### PR TITLE
Adding support for scaling mode: MaintenanceWindow

### LINKED ISSUE #12: [Feature] Support maintenance window especially for scaling down
**Feature (What you would like to be added):**
We already have the [`Duration`](https://github.com/gardener/hvpa-controller/blob/master/pkg/apis/autoscaling/v1alpha1/hvpa_types.go#L89) field in `ScaleStabilization` struct used separately for both scale up and scale down to specify the minimum time between successive scaling by HVPA.

As an alternative and/or complement to this, we should support an additional `MaintenanceWindow` field to indicate that the corresponding scaling (typically, scale down) should be done only during that time window.

**Motivation (Why is this needed?):**
If the target of HVPA does not support non-disruptive scale-down then we can at least support scale down during a user-specified maintenance window to avoid unexpected downtimes due to scaling operations.

**Approach/Hint to the implement solution (optional):**
1. The recommendations from VPA and HPA during the specified maintenance window may not be right values for scaling. It might be better to be conservative.
For example, record (somewhere in the status section of the HVPA resource) the maximum recommendations from VPA and HPA during the period out of the maintenance window and then apply this value during the maintenance window.

1. Typically, only one of `Duration` or `MaintenanceWindow` would be specified.
But if both are specified, then it could be taken that `Duration` is applicable when scaling is done multiple times during the speficied maintenance window.

### PR DESCRIPTION
**Pull request by ggaurav10**
**What this PR does / why we need it**:
If the target of HVPA does not support non-disruptive scale-down then we can at least support scale down during a user-specified maintenance window to avoid unexpected downtimes due to scaling operations.

**Which issue(s) this PR fixes**:
Fixes #12 

**Special notes for your reviewer**:
I have done some testing, and still going through some test scenarios which I can add to the test cases once I am through. Raising this PR so that review can start in parallel.

**Release note**:

```improvement operator
HVPA now supports setting scale mode to `MaintenanceWindow`. When this mode is set, scaling happens only during user-defined maintenance time window.
Currently, this mode is supported only for vertical scaling. Support for horizontal scaling is dependent on implementation of `scale` subresource in HVPA
```

### CODE REVIEW

File: controllers/hvpa_controller.go [hunk 1]
```diff
@@ -338,7 +339,17 @@ func (r *HvpaReconciler) reconcileHpa(hvpa *autoscalingv1alpha1.Hvpa) (*autoscal
 		return nil, err
 	}
 
-	updatePolicy := hvpa.Spec.Hpa.UpdatePolicy.UpdateMode
+	toDeploy := hvpa.Spec.Hpa.Deploy
+
+	upUpdatePolicy, downUpdatePolicy := autoscalingv1alpha1.UpdateModeAuto, autoscalingv1alpha1.UpdateModeAuto
```
Comment (amshuman-kr) [inline controllers/hvpa_controller.go]: This is only for the time being, right? Until the `Scale` sub-resource has been implemented.

File: controllers/hvpa_controller_test.go [hunk 2]
```diff
@@ -140,6 +179,262 @@ var _ = Describe("#TestReconcile", func() {
...
+					desiredReplicas: 3,
+					resourceChange:  false,
+					scaleOutLimited: false,
+					resources:       unscaled,
+					blockedReasons: []autoscalingv1alpha1.BlockingReason{
+						autoscalingv1alpha1.BlockingReasonUpdatePolicy,
+					},
+				},
+			}),
+			Entry("UpdateMode Auto, Proportional scaling", &data{
```
Comment (amshuman-kr) [inline controllers/hvpa_controller_test.go]: Can we have some more cases for the limit scaling to cover other cases?

File: utils/timewindow.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+type MaintenanceTime struct {
+	hour   int
+	minute int
+	second int
+}
+
+// NewMaintenanceTime returns a maintenance time structure based on the given hour, minute, and second.
+func NewMaintenanceTime(hour, minute, second int) *MaintenanceTime {
+	if hour >= 24 {
+		panic(fmt.Sprintf("invalid hour %d", hour))
```
Comment (amshuman-kr) [inline utils/timewindow.go]: Is `panic` necessary? Won't returning error do? Some misconfigured CRD should not cause the controller to panic, no?

File: utils/timewindow.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+	}
+	if minute >= 60 {
+		panic(fmt.Sprintf("invalid minute %d", minute))
```
Comment (amshuman-kr) [inline utils/timewindow.go]: Same as [above](https://github.com/gardener/hvpa-controller/pull/50/files#diff-55686f78c3ca861b865b1cbd5b4604f3R34)

File: utils/timewindow.go [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+	}
+	if second >= 60 {
+		panic(fmt.Sprintf("invalid second %d", second))
```
Comment (amshuman-kr) [inline utils/timewindow.go]: Same as [above](https://github.com/gardener/hvpa-controller/pull/50/files#diff-55686f78c3ca861b865b1cbd5b4604f3R34)

File: utils/timewindow.go [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+// Second returns the second of the maintenance time.
+func (m *MaintenanceTime) Second() int {
+	return m.second
+}
+
+// Add adds hour, minute and second to <m> and returns a new maintenance time.
+func (m *MaintenanceTime) Add(hour, minute, second int) *MaintenanceTime {
+	t := m.zeroTime().Add(time.Duration(hour)*time.Hour + time.Duration(minute)*time.Minute + time.Duration(second)*time.Second)
```
Comment (amshuman-kr) [inline utils/timewindow.go]: What's the reasoning behind `zeroTime` here?

**Review by amshuman-kr — CHANGES_REQUESTED** [review]
Excellent PR! Some minor suggestions.

File: utils/timewindow.go (hunk 3 above)
Comment (ggaurav10) [reply utils/timewindow.go]: This is not supposed to happen even if user misconfigures the CRD

**Review by amshuman-kr — APPROVED** [review]
Thanks for all the test cases! LGTM

**Review by amshuman-kr — APPROVED** [review]
LGTM

**Pull request merged**